File: src/main/java/Day2.kt
import java.util.regex.Pattern

class Day2: Day<Array<IntArray>> {
    override fun part1(input: Array<IntArray>): Any {
        return perform(input, { highLowDiff(it) })
    }

    override fun part2(input: Array<IntArray>): Any {
        return perform(input, { divisible(it) })
    }

    private fun divisible(numbers: IntArray): Int {
        numbers.forEach { a ->
            numbers.forEach { b ->
                if (a != b && a % b == 0) {
                    return a / b
                }
            }
        }
        return 0
    }

    override fun parse(text: String): Array<IntArray> {
        return text.split('\n').map { s -> s.trim().split(Pattern.compile("\\s+"))
                .map { it.trim().toInt() }.toIntArray() }.toTypedArray()
    }

    fun highLowDiff(numbers: IntArray): Int {
        return numbers.max()!! - numbers.min()!!
    }

    fun perform(values: Array<IntArray>, operation: (IntArray) -> Int): Int {
        return values.fold(0, { sum, next -> sum + operation.invoke(next) })
    }

}

File: 2019-typescript/src/Day2.ts
import Day from "./Day";
import IntCodeComputer from "./IntCodeComputer"

class Day2 implements Day<number[]> {

  parse(text: string): number[] {
    return text.split(',').map((x: string) => parseInt(x, 10));
  }

  part1(input: number[]) {
    input[1] = 12;
    input[2] = 2;
    return this.runProgram(input.slice());
  }

  part2(input: number[]) {
    for (let a = 0; a <= 99; a++) {
      for (let b = 0; b <= 99; b++) {
        let copy = input.slice()
        copy[1] = a;
        copy[2] = b;
        if (this.runProgram(copy) === 19690720) {
          return 100 * a + b;
        }
      }
    }
    throw new Error("Not found");
  }

  runProgram(values: number[]) {
    let computer = new IntCodeComputer()
    computer.runProgram(values);
    return computer.tape[0];
  }
}

export default Day2;
